app-ios/tutanota/Sources/Alarms/AlarmModel.swift

```diff
@@ -2,235 +2,169 @@ import Foundation
-    localTimeZone: TimeZone
-  ) -> LazyEventSequence {
-    var cal = Calendar.current
-    let calendarUnit = calendarUnit(for: repeatRule.frequency)
-
-    let isAllDayEvent = isAllDayEvent(startTime: eventStart, endTime: eventEnd)
-    let calcEventStart = isAllDayEvent ? allDayLocalDate(fromUTCDate: eventStart) : eventStart
-    let endDate: Date?
-    switch repeatRule.endCondition {
-    case let .untilDate(valueDate):
-      if isAllDayEvent {
-        endDate = allDayLocalDate(fromUTCDate: valueDate)
-      } else {
-        endDate = valueDate
-      }
-    default:
-      endDate = nil
-    }
-
-    cal.timeZone = isAllDayEvent ? localTimeZone : TimeZone(identifier: repeatRule.timeZone) ?? localTimeZone
-
-    return LazyEventSequence(calcEventStart: calcEventStart, endDate: endDate, repeatRule: repeatRule, cal: cal, calendarComponent: calendarUnit)
-  }
-
-  static func alarmTime(trigger: AlarmInterval, eventTime: Date) -> Date {
-    let cal = Calendar.current
-    switch trigger.unit {
-    case .minute:
-      return cal.date(byAdding: .minute, value: -trigger.value, to: eventTime)!
-    case .hour:
-      return cal.date(byAdding: .hour, value: -trigger.value, to: eventTime)!
-    case .day:
-      return cal.date(byAdding: .day, value: -trigger.value, to: eventTime)!
-    case .week:
-      return cal.date(byAdding: .weekOfYear, value: -trigger.value, to: eventTime)!
-    }
-  }
+	private let dateProvider: DateProvider
+
+	init(dateProvider: DateProvider) { self.dateProvider = dateProvider }
+
+	func futureOccurrences(acrossAlarms alarms: [AlarmNotification], upToForEach: Int, upToOverall: Int) -> any BidirectionalCollection<AlarmOccurence> {
+		var occurrences = [AlarmOccurence]()
+
+		for alarm in alarms {
+			let a = prefix(self.futureOccurrences(ofAlarm: alarm), upToForEach)
+			occurrences += a
+		}
+
+		occurrences.sort(by: { $0.eventOccurrenceTime < $1.eventOccurrenceTime })
+		return occurrences.prefix(upToOverall)
+	}
+
+	func futureOccurrences(ofAlarm alarm: AlarmNotification) -> any Sequence<AlarmOccurence> {
+		if let repeatRule = alarm.repeatRule {
+			return self.futureOccurences(ofAlarm: alarm, withRepeatRule: repeatRule)
+		} else {
+			let singleOcurrence = AlarmOccurence(occurrenceNumber: 0, eventOccurrenceTime: alarm.eventStart, alarm: alarm)
+			if shouldScheduleAlarmAt(ocurrenceTime: singleOcurrence.alarmOccurenceTime()) { return [singleOcurrence] } else { return [] }
+		}
+	}
+
+	private func futureOccurences(ofAlarm alarm: AlarmNotification, withRepeatRule: RepeatRule) -> some Sequence<AlarmOccurence> {
+		let occurencesAfterNow = occurencesOfRepeatingEvent(
+			eventStart: alarm.eventStart, eventEnd: alarm.eventEnd, repeatRule: withRepeatRule, localTimeZone: TimeZone.current
+		).lazy  // trying to optimize it: do not calculate alarm occurence if event occurence itself is in the past
+			.filter { self.shouldScheduleAlarmAt(ocurrenceTime: $0.occurenceDate) }.map { occurrence in
```

Comment:
uh I dislike this chaining style. Can we use `lineBreakAroundMultilineExpressionChainComponents` ? What do you think?

Reply:
Looks better now, I think

Reply:
much better now, thank you!